wgsl/extract-grammar.py

```diff
@@ -502,17 +743,196 @@ def read_spec(options):
+                    scanner_i += scanner_end[-1]
         scanner_i += 1
 
-    result[scanner_rule.name()]["_comment"] = [["`/\\/\\/.*/`"]]
+    bnfdialect_data = ""
+
+    # Read the contents of the bnfdialect file
+    with open(options.syntax_filename, "r") as file:
+        bnfdialect_data = file.read()
+
+    syntax_dict = {}
+    current_rule = None
+    for line in bnfdialect_data.splitlines():
+        comment = ""
+        if "//" in line:
+            line, comment = line.split("//", 1)
+        line = line.strip() # Either blankspace is insignificant
+        comment = comment.rstrip() # Right blankspace is insignificant
+        if len(line) == 0:
+            continue
+        elif current_rule is None:
+            if line.endswith(":"):
+                current_rule = line[:-1].strip()
+                syntax_dict[current_rule] = {
+                    "type": "sequence",
+                    "value": []
+                }
+        else:
+            if line.startswith(";"):
+                current_rule = None
+            else:
+                if line.startswith("|"):
+                    syntax_dict[current_rule]["type"] = "choice"
+                    line = line[1:].strip()
+                tokens = [i for i in line.split(" ") if len(i) > 0] # Strip blankspace
+                def build_production(token_j):
+                    """
+                    Takes token index and returns token steps and a (sub)production
+                    """
+                    if tokens[token_j].startswith("/"):
+                        return (1, {
+                            "type": "pattern",
+                            "value": tokens[token_j]
+                        })
+                    if tokens[token_j].startswith("'"):
+                        return (1, {
+                            "type": "token",
+                            "value": tokens[token_j]
+                        })
+                    if tokens[token_j].startswith("?"):
+                        value = build_production(token_j - 1)
+                        return (1 + value[0], {
+                            "type": "match_0_1",
+                            "value": [value[1]]
+                        })
+                    if tokens[token_j].startswith("+"):
+                        value = build_production(token_j - 1)
+                        return (1 + value[0], {
+                            "type": "match_1_n",
+                            "value": [value[1]]
+                        })
+                    if tokens[token_j].startswith("*"):
+                        value = build_production(token_j - 1)
+                        return (1 + value[0], {
+                            "type": "match_0_n",
+                            "value": [value[1]]
+                        })
+                    if tokens[token_j].startswith("|"):
+                        return (1, {
+                            "type": "__MAKE_CHOICE",
```

Comment:
It seems like this should have to backtrack and make a sequence of the preceding tokens, at least until the previous |  or  (


For example if I had:

    a_rule : 
      lhs_expression ( 'atomic' '=' | compound_assignment )  expression
    ;

Then when this parser hit the   |    it would not know to backtrack to make a sequence from 'atomic' '='.

This bad case does not happen in syntax.bnf. The closest is the assignment operator.

```
assignment_statement :
  lhs_expression ( '=' | compound_assignment_operator ) expression
| '_' '=' expression
; 
```


So, this is a bug waiting to be triggered.


I think the recursive approach had in my prototype was clearer and more robust.

Reply:
I'm not sure if that case would be a bug since the equivalence check between the reconstructed source and provided source can't match such a scenario. I've thought that we'd make such rules require parentheses, as a glance reading of things like ( expr '=' | '+= ) sometimes come across as if ( expr ( '=' | '+=' ) ) while being the other way around.

Reply:
I did an experiment changing to this:

```
assignment_statement :
  lhs_expression ( '|' '=' | compound_assignment_operator ) expression
| '_' '=' expression
; 
```

I verified that yes, it is mis-parsed (in my view), but  also that the differ catches the bug.

```
ERROR: Syntax source should match reproduction for styling and language
--- 

+++ 

@@ -316,7 +316,7 @@

 ;
 
 assignment_statement :
-lhs_expression ( ':' '=' | compound_assignment_operator ) expression
+lhs_expression ( ':' | '=' | compound_assignment_operator ) expression
 | '_' '=' expression
 ;
 
make: *** [validate-examples] Error 1
```

So the bad case I imagined would get caught.  I can live with this.

I also believe a later enhancement can fix the issue, if we spent more time reworking the parsing.